distributed/worker.py

```diff
@@ -1444,56 +1460,50 @@ def add_task(
         actor=False,
         **kwargs2,
     ):
+        runspec = SerializedTask(function, args, kwargs, task)
         try:
             if key in self.tasks:
                 ts = self.tasks[key]
+                ts.runspec = runspec
                 if ts.state == "memory":
                     assert key in self.data or key in self.actors
                     logger.debug(
                         "Asked to compute pre-existing result: %s: %s", key, ts.state
                     )
                     self.send_task_state_to_scheduler(ts)
                     return
-                if ts.state in IN_PLAY:
-                    return
-                if ts.state == "erred":
-                    ts.exception = None
-                    ts.traceback = None
-                else:
-                    ts.state = "waiting"
             else:
                 self.log.append((key, "new"))
-                self.tasks[key] = ts = TaskState(
-                    key=key, runspec=SerializedTask(function, args, kwargs, task)
-                )
+                self.tasks[key] = ts = TaskState(key=key)
                 ts.state = "waiting"
 
-            if priority is not None:
-                priority = tuple(priority) + (self.generation,)
-                self.generation -= 1
-
             if actor:
                 self.actors[ts.key] = None
 
             ts.priority = priority
             ts.duration = duration
+            ts.runspec = runspec
             if resource_restrictions:
                 ts.resource_restrictions = resource_restrictions
 
             who_has = who_has or {}
 
             for dependency, workers in who_has.items():
-                assert workers
+                if self.address in workers:
+                    logger.debug(
+                        f"Who has claims {self.name} would own data of {dependency} but this is untrue."
+                    )
+                    # workers = tuple(w_addr for w_addr in workers if w_addr != self.address)
                 if dependency not in self.tasks:
                     self.tasks[dependency] = dep_ts = TaskState(key=dependency)
-                    dep_ts.state = (
-                        "waiting" if dependency not in self.data else "memory"
-                    )
-
-                dep_ts = self.tasks[dependency]
-                self.log.append((dependency, "new-dep", dep_ts.state))
+                    dep_ts.state = "waiting"
```

Comment:
I'm certainly not sure that my approach was the correct one, but if we call `release_key` on a task that has any dependents, then the key will end up popped out of `self.tasks` without also clearing the corresponding item out of `self.data` -- so if we set the state to `waiting` here without checking for presence in `self.data` it can fail the check in `validate_task_waiting`

Reply:
Sure, will check this section again. In general I'm not very happy with the approach of modifying the state attribute manually and maybe this should also by done via a proper transition step. tbh, I haven't fully understood your comment, yet, but if multiple tasks / dependencies must be adjusted this sounds like something which should be handled by a transition method anyhow I suppose

Reply:
My comment was very unclear (having just re-read it).

It is possible at this point to have a task whose result is present in `self.data` even though there is no corresponding `TaskState` object in `self.tasks`.  If that is the case, then the correct `state` is `memory`.  If we blanket set it to `waiting` it will fail the assertions in `validate_task_waiting`.

Reply:
How can data end up in memory without it being a task? (other than manual manipulation of the dict)

Reply:
I indeed am seeing this error case you are describing but I'm wondering if this is an allowed state altogether. Even when effectively setting data _manually_ (i.e. using the respective API) a `TaskState` object is created if it doesn't exist.

https://github.com/dask/distributed/blob/49e19c010b101c629691c73d62caee7140bb92d9/distributed/worker.py#L1387-L1398

 I was working under the assumption that `data` is always a subset of `tasks`

Reply:
I can find a section of code which doesn't delete data even though the key is released if there are dependents of the task. 
https://github.com/dask/distributed/blob/49e19c010b101c629691c73d62caee7140bb92d9/distributed/worker.py#L2255-L2268
I can somehow understand this reasoning but am wondering how the data is then cleaned up once it is no longer required since `release_key` is the only interface on how other servers can interact with the tasks and the data.

Even `delete_data` itself uses this, see https://github.com/dask/distributed/blob/49e19c010b101c629691c73d62caee7140bb92d9/distributed/worker.py#L1405-L1412

Doesn't this introduce memory leakage?

Reply:
> Doesn't this introduce memory leakage?

Probably?  We could try removing that case (in `release_key`) and see what breaks.

Reply:
Ok I think I found what mechanism allowed this to not be discovered / function properly. There is something like a reference counting ongoing, i.e. the dependents and dependencies are all removed from the task states individually, e.g. 
https://github.com/dask/distributed/blob/49e19c010b101c629691c73d62caee7140bb92d9/distributed/worker.py#L1734-L1737
and 
https://github.com/dask/distributed/blob/49e19c010b101c629691c73d62caee7140bb92d9/distributed/worker.py#L2271-L2277

This allows for a cleanup of the data even though the data is no longer tracked as a task. I deleted some of the code since I considered dependencies/dependents as a immutable property of a task and not a runtime dependent one. That confused me quite a bit. I don't necessarily agree with the implementation right now but I'll restore everything as it was to reduce the number of changes.

Reply:
> That confused me quite a bit. I don't necessarily agree with the implementation right now but I'll restore everything as it was to reduce the number of changes.

I'm definitely game to sync up at some point and sketch out a cleaner implementation for this.

Reply:
> I'm definitely game to sync up at some point and sketch out a cleaner implementation for this.

Sounds good. Once this PR is finally stable, I'll open an issue and summarise the points I've seen where I think we can improve and we can sync up on the topic.

Reply:
I updated the PR with my current state. Unfortunately, reinstating the full dependency/release/refcounting as described above seemed to make things worse :(

Reply:
> Doesn't this introduce memory leakage?

Seeing the question about memory leak, I'm curious if this could be related to https://github.com/dask/distributed/issues/3898.

Reply:
If my argumentation about the memory leak was true, the leak would happen on the worker since the worker would not release the data anymore. From my understanding, #3898 is about a potential leak on the scheduler.
Also, the leak I was referring to does not exist and I was mistaken. The worker applies some refcounting by tracking the `dependents` of each task and releases the data properly once all dependents are gone, i.e. the data is no longer needed. I missed this mechanism since it's a bit spread out in this module but from my understanding it works as intended.